File: src/com/jupiter/ganymede/property/Property.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.ganymede.property;

import com.jupiter.ganymede.event.Event;

/**
 *
 * @author NathanT
 * @param <T>
 */
public class Property<T> {

    // Events
    public final Event<PropertyChangedArgs<T>> changed = new Event<>();
    
    
    // Fields
    private T value;
    private PropertyBinding binding;


    // Initialization
    /**
     * Initializes a Property with initial value null.
     */
    public Property() {
        this(null);
    }

    /**
     * Initializes a property with the given initial value.
     *
     * @param initialValue
     */
    public Property(T initialValue) {
        this.value = initialValue;
    }


    // Public Methods
    /**
     * Fetches the current value of the Property.
     *
     * @return
     */
    public T get() {
        return this.value;
    }

    /**
     * Sets the value of the Property.
     *
     * @param value
     */
    public void set(T value) {
        T oldValue = this.value;
        this.value = value;

        if (!oldValue.equals(this.value)) {
            this.changed.dispatch(new PropertyChangedArgs<>(this, oldValue, this.value));
        }
    }

    /**
     * Binds this property to another so that it will reflect the other Property's value.
     *
     * @param other
     */
    public void bind(Property<T> other) {
        this.bind(new PropertyBinding(() -> other.get(), other));
    }

    /**
     * Binds this property to an arbitrary expression.
     *
     * @param binding
     * @param dependencies The other properties that the binding is dependent on. The binding will be recalculated
     * whenever one of these properties is changed.
     *
     */
    public void bind(PropertyBinding<T> binding) {
        this.unbind();
        this.binding = binding;
        this.binding.bind(this);
    }

    /**
     * Removes the current binding from the property.
     */
    public void unbind() {
        if (this.binding != null) {
            this.binding.unbind();
            this.binding = null;
        }
    }
    
    
    // Inner Classes
    public static class PropertyChangedArgs<T> {
        
        // Fields
        public final Property<T> property;
        public final T oldValue;
        public final T newValue;
        
        
        // Initialization
        public PropertyChangedArgs(Property<T> property, T oldValue, T newValue) {
            this.property = property;
            this.oldValue = oldValue;
            this.newValue = newValue;
        }
        
    }

}

File: src/com/jupiter/ganymede/property/Property.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.ganymede.property

import com.jupiter.ganymede.event.Event
import com.jupiter.ganymede.event.Listener

/**

 * @author NathanT
 * *
 * @param
 */
class Property<T>(initialValue: T) {

    // Events
    private val changed = Event<PropertyChangedArgs<T>>()


    // Fields
    private var value: T = initialValue
    private var binding: PropertyBinding<T>? = null


    fun get(): T = this.value

    /**
     * Sets the value of the Property.

     * @param value
     */
    fun set(value: T) {
        val oldValue: T = this.value
        this.value = value

        if ((oldValue != null && oldValue != this.value) || (oldValue == null && this.value != null)) {
            this.changed.dispatch(PropertyChangedArgs(
                    property = this,
                    newValue = value,
                    oldValue = oldValue))
        }
    }

    fun addPropertyChangedListener(listener: (PropertyChangedArgs<T>) -> Unit): Boolean {
        return this.changed.addListener(listener)
    }

    fun removePropertyChangedListener(listener: (PropertyChangedArgs<T>) -> Unit): Boolean {
        return this.changed.removeListener(listener)
    }

    /**
     * Binds this property to another so that it will reflect the other Property's value.

     * @param other
     */
    fun bind(other: Property<T>) {
        this.bind(PropertyBinding({ other.get() }))
    }

    /**
     * Binds this property to an arbitrary expression.

     * @param binding
     */
    fun bind(binding: PropertyBinding<T>) {
        this.unbind()
        this.binding = binding
        this.binding?.bind(this)
    }

    /**
     * Removes the current binding from the property.
     */
    fun unbind() {
        if (this.binding != null) {
            this.binding?.unbind()
            this.binding = null
        }
    }


    // Inner Classes
    data class PropertyChangedArgs<T>(val property: Property<T>, val oldValue: T, val newValue: T)

    interface PropertyListener<T> : Listener<PropertyChangedArgs<T>>
}